### ISSUE TITLE
No way to catch error thrown by WebSocket.addError()

### ISSUE THREAD

**Issue by evanweible-wf** [dart-lang/sdk#24278]
Consider the following simple usage of the dart:io WebSocket class:

```dart
import 'dart:async';
import 'dart:io';

main() async {
  // Connect to a web socket.
  WebSocket socket = await WebSocket.connect('ws://echo.websocket.org');

  // Setup listening.
  Completer done = new Completer();
  socket.listen((message) {
    print('message: $message');
  }, onError: (error) {
    print('error: $error');
  }, onDone: () {
    print('socket closed.');
    done.complete();
  }, cancelOnError: true);

  // Add message, and then an error.
  socket.add('echo!');
  socket.addError(new Exception('error!'));

  // Wait for the socket to close.
  await done.future;
}
```

When you add an error to the websocket sink, it eventually gets added to the StreamConsumer tied to the underlying socket, which results in a completer being completed with the error. This happens in an async zone, and the exception is thrown, but cannot be caught.

On one hand, this is fine, because calling addError on a WebSocket instance should result in some sort of exception, but it makes it impossible (from what I can tell?) to wrap the `WebSocket` class and handle that error. I'm attempting to do so currently, and would like to be able to catch that error for testing purposes. Should this be possible?

**Comment by sgjesse** [comment]
@lrhn 

You can use the `done` future on the WebSocket object to get that error, e.g.:

```
import 'dart:async';
import 'dart:io';

main() async {
  // Connect to a web socket.
  WebSocket socket = await WebSocket.connect('ws://echo.websocket.org');

  // Setup listening.
  socket.listen((message) {
    print('message: $message');
  }, onError: (error) {
    print('error: $error');
  }, onDone: () {
    print('socket closed.');
  }, cancelOnError: true);

  // Add message, and then an error.
  socket.add('echo!');
  socket.addError(new Exception('error!'));

  // Wait for the socket to close.
  try {
    await socket.done;
    print('WebSocket donw');
  } catch (error) {
    print('WebScoket done with error $error');
  }
}
```

However there was still a bug causing the subscription not to cancel. I have a fix for that in https://codereview.chromium.org/1304243005.

**Comment by evanweible-wf** [comment]
Great! I guess that makes sense since `done` is tied to the WebSocket's sink. And thanks for the quick fix on the other part!

**Comment by sgjesse** [comment]
Fixed in 373dd4b33c1efd52ef0dc0ad2a5249263b6b506d
